File: static/js/alerts.js
// API Base URL
const API_BASE = '/api';

// Current filters
let currentFilters = {
    status: 'active',
    heatExchanger: 'all',
    severity: 'all'
};

// WebSocket connection
let ws = null;

// Load alerts on page load
document.addEventListener('DOMContentLoaded', () => {
    checkUserRole();
    loadHeatExchangers();
    loadAlerts();
    setupFilters();
    connectWebSocket();
    
    // Refresh every 30 seconds
    setInterval(loadAlerts, 30000);
});

// Check user role and display navigation
async function checkUserRole() {
    try {
        const response = await fetch('/api/auth/me');
        if (response.ok) {
            const user = await response.json();
            
            // Show user info and logout
            document.getElementById('currentUser').textContent = `👤 ${user.username}`;
            document.getElementById('currentUser').style.display = 'inline';
            document.getElementById('logoutBtn').style.display = 'inline-block';
            
            if (user.is_admin) {
                // Show admin links
                document.getElementById('addHeatExchangerLink').style.display = 'inline-block';
                document.getElementById('usersLink').style.display = 'inline-block';
                document.getElementById('settingsLink').style.display = 'inline-block';
            }
            
            // Setup logout
            document.getElementById('logoutBtn').addEventListener('click', async () => {
                await fetch('/api/auth/logout', { method: 'POST' });
                window.location.href = '/login';
            });
        } else {
            // Show login button
            document.getElementById('loginLink').style.display = 'inline-block';
        }
    } catch (error) {
        console.log('Not authenticated:', error);
        document.getElementById('loginLink').style.display = 'inline-block';
    }
}

function connectWebSocket() {
    const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
    ws = new WebSocket(`${protocol}//${window.location.host}/ws`);
    
    ws.onmessage = (event) => {
        const data = JSON.parse(event.data);
        if (data.type === 'new_alert') {
            console.log('New alert received via WebSocket:', data);
            loadAlerts(); // Reload alerts
        }
    };
    
    ws.onclose = () => {
        console.log('WebSocket closed, reconnecting...');
        setTimeout(connectWebSocket, 5000);
    };
}

async function loadHeatExchangers() {
    try {
        const response = await fetch(`${API_BASE}/heat-exchangers`);
        const exchangers = await response.json();
        
        const select = document.getElementById('filterHeatExchanger');
        exchangers.forEach(he => {
            const option = document.createElement('option');
            option.value = he.id;
            option.textContent = he.name;
            select.appendChild(option);
        });
    } catch (error) {
        console.error('Error loading heat exchangers:', error);
    }
}

async function loadAlerts() {
    try {
        // Build query parameters
        const params = new URLSearchParams();
        
        if (currentFilters.status === 'active') {
            params.append('acknowledged', 'false');
            params.append('resolved', 'false');
        } else if (currentFilters.status === 'acknowledged') {
            params.append('acknowledged', 'true');
            params.append('resolved', 'false');
        } else if (currentFilters.status === 'resolved') {
            params.append('resolved', 'true');
        }
        
        if (currentFilters.heatExchanger !== 'all') {
            params.append('heat_exchanger_id', currentFilters.heatExchanger);
        }
        
        if (currentFilters.severity !== 'all') {
            params.append('severity', currentFilters.severity);
        }
        
        const response = await fetch(`${API_BASE}/alerts?${params.toString()}`);
        const alerts = await response.json();
        
        // Update stats
        updateStats(alerts);
        
        // Render alerts
        renderAlerts(alerts);
    } catch (error) {
        console.error('Error loading alerts:', error);
    }
}

async function updateStats(alerts) {
    try {
        // Get counts
        const totalResponse = await fetch(`${API_BASE}/alerts/count`);
        const totalData = await totalResponse.json();
        
        const activeResponse = await fetch(`${API_BASE}/alerts/count?acknowledged=false&resolved=false`);
        const activeData = await activeResponse.json();
        
        const ackResponse = await fetch(`${API_BASE}/alerts/count?acknowledged=true&resolved=false`);
        const ackData = await ackResponse.json();
        
        const resolvedResponse = await fetch(`${API_BASE}/alerts/count?resolved=true`);
        const resolvedData = await resolvedResponse.json();
        
        document.getElementById('totalAlerts').textContent = totalData.count;
        document.getElementById('activeAlerts').textContent = activeData.count;
        document.getElementById('acknowledgedAlerts').textContent = ackData.count;
        document.getElementById('resolvedAlerts').textContent = resolvedData.count;
    } catch (error) {
        console.error('Error updating stats:', error);
    }
}

function renderAlerts(alerts) {
    const container = document.getElementById('alertsContainer');
    const noAlerts = document.getElementById('noAlerts');
    
    if (alerts.length === 0) {
        container.innerHTML = '';
        noAlerts.style.display = 'block';
        return;
    }
    
    noAlerts.style.display = 'none';
    container.innerHTML = alerts.map(alert => createAlertCard(alert)).join('');
}

function createAlertCard(alert) {
    const createdDate = new Date(alert.created_at).toLocaleString();
    const ackDate = alert.acknowledged_at ? new Date(alert.acknowledged_at).toLocaleString() : null;
    const resolvedDate = alert.resolved_at ? new Date(alert.resolved_at).toLocaleString() : null;
    
    let statusClass = '';
    if (alert.resolved) statusClass = 'resolved';
    else if (alert.acknowledged) statusClass = 'acknowledged';
    
    let badges = `<span class="badge badge-${alert.severity}">${alert.severity}</span>`;
    if (alert.acknowledged) badges += '<span class="badge badge-acknowledged">Acknowledged</span>';
    if (alert.resolved) badges += '<span class="badge badge-resolved">Resolved</span>';
    
    let actions = '';
    if (!alert.resolved) {
        if (!alert.acknowledged) {
            actions = `
                <button class="btn btn-small" onclick="acknowledgeAlert(${alert.id})">
                    ✓ Acknowledge
                </button>
                <button class="btn btn-small" onclick="resolveAlert(${alert.id})">
                    ✓✓ Resolve
                </button>
            `;
        } else {
            actions = `
                <button class="btn btn-small" onclick="resolveAlert(${alert.id})">
                    ✓✓ Resolve
                </button>
            `;
        }
        actions += `
            <button class="btn btn-small btn-secondary" onclick="toggleCommentForm(${alert.id})">
                💬 Add Comment
            </button>
        `;
    }
    
    let metaInfo = '';
    if (alert.acknowledged_by) {
        metaInfo += `<div>Acknowledged by ${alert.acknowledged_by} on ${ackDate}</div>`;
    }
    if (alert.resolved_by) {
        metaInfo += `<div>Resolved by ${alert.resolved_by} on ${resolvedDate}</div>`;
    }
    
    return `
        <div class="card alert-card ${alert.severity} ${statusClass}">
            <div class="alert-header">
                <div>
                    <h3 class="alert-title">${alert.title}</h3>
                    <div class="alert-meta">
                        <strong>${alert.heat_exchanger_name}</strong> • ${createdDate}
                    </div>
                    ${metaInfo ? `<div class="alert-meta" style="margin-top: 0.5rem;">${metaInfo}</div>` : ''}
                </div>
                <div class="alert-badges">
                    ${badges}
                </div>
            </div>
            
            ${alert.description ? `<p>${alert.description}</p>` : ''}
            
            <div class="alert-details">
                ${alert.pump_name ? `
                    <div class="detail-item">
                        <span class="detail-label">Pump</span>
                        <span class="detail-value">${alert.pump_name}</span>
                    </div>
                ` : ''}
                ${alert.flow_rate !== null ? `
                    <div class="detail-item">
                        <span class="detail-label">Flow Rate</span>
                        <span class="detail-value" style="color: var(--danger-color); font-weight: 600;">
                            ${alert.flow_rate.toFixed(2)} L/min
                        </span>
                    </div>
                ` : ''}
                ${alert.threshold !== null ? `
                    <div class="detail-item">
                        <span class="detail-label">Threshold</span>
                        <span class="detail-value">${alert.threshold.toFixed(2)} L/min</span>
                    </div>
                ` : ''}
            </div>
            
            ${alert.comments ? `
                <div class="alert-comments">
                    <strong>Comments:</strong><br>
                    ${alert.comments}
                </div>
            ` : ''}
            
            <div class="alert-actions">
                ${actions}
            </div>
            
            <div id="commentForm${alert.id}" class="comment-form" style="display: none;">
                <input 
                    type="text" 
                    id="commentInput${alert.id}" 
                    placeholder="Add a comment..."
                    onkeypress="if(event.key==='Enter') addComment(${alert.id})"
                >
                <button class="btn btn-small" onclick="addComment(${alert.id})">Post</button>
                <button class="btn btn-small btn-secondary" onclick="toggleCommentForm(${alert.id})">Cancel</button>
            </div>
        </div>
    `;
}

async function acknowledgeAlert(alertId) {
    try {
        const response = await fetch(`${API_BASE}/alerts/${alertId}/acknowledge`, {
            method: 'PUT',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({})
        });
        
        if (!response.ok) throw new Error('Failed to acknowledge alert');
        
        loadAlerts(); // Reload
    } catch (error) {
        console.error('Error acknowledging alert:', error);
        alert('Failed to acknowledge alert');
    }
}

async function resolveAlert(alertId) {
    try {
        const response = await fetch(`${API_BASE}/alerts/${alertId}/resolve`, {
            method: 'PUT',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({})
        });
        
        if (!response.ok) throw new Error('Failed to resolve alert');
        
        loadAlerts(); // Reload
    } catch (error) {
        console.error('Error resolving alert:', error);
        alert('Failed to resolve alert');
    }
}

function toggleCommentForm(alertId) {
    const form = document.getElementById(`commentForm${alertId}`);
    form.style.display = form.style.display === 'none' ? 'flex' : 'none';
}

async function addComment(alertId) {
    const input = document.getElementById(`commentInput${alertId}`);
    const comment = input.value.trim();
    
    if (!comment) return;
    
    try {
        const response = await fetch(`${API_BASE}/alerts/${alertId}/comment`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ comments: comment })
        });
        
        if (!response.ok) throw new Error('Failed to add comment');
        
        input.value = '';
        toggleCommentForm(alertId);
        loadAlerts(); // Reload
    } catch (error) {
        console.error('Error adding comment:', error);
        alert('Failed to add comment');
    }
}

function setupFilters() {
    document.getElementById('filterStatus').addEventListener('change', (e) => {
        currentFilters.status = e.target.value;
        loadAlerts();
    });
    
    document.getElementById('filterHeatExchanger').addEventListener('change', (e) => {
        currentFilters.heatExchanger = e.target.value;
        loadAlerts();
    });
    
    document.getElementById('filterSeverity').addEventListener('change', (e) => {
        currentFilters.severity = e.target.value;
        loadAlerts();
    });
}

File: app/routers/alerts.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, desc, and_
from datetime import datetime
from typing import List, Optional

from app.database import get_session
from app.models.alert import Alert, AlertResponse, AlertAcknowledge, AlertResolve, AlertComment
from app.models.heat_exchanger import HeatExchanger
from app.models.user import User
from app.routers.auth import get_current_user

router = APIRouter(prefix="/api/alerts", tags=["alerts"])


@router.get("/", response_model=List[AlertResponse])
async def get_alerts(
    heat_exchanger_id: Optional[int] = Query(None),
    acknowledged: Optional[bool] = Query(None),
    resolved: Optional[bool] = Query(None),
    severity: Optional[str] = Query(None),
    limit: int = Query(100, le=1000),
    db: AsyncSession = Depends(get_session),
    current_user: User = Depends(get_current_user)
):
    """Get all alerts with optional filters"""
    query = select(Alert, HeatExchanger.name).join(
        HeatExchanger, Alert.heat_exchanger_id == HeatExchanger.id
    )
    
    # Apply filters
    filters = []
    if heat_exchanger_id is not None:
        filters.append(Alert.heat_exchanger_id == heat_exchanger_id)
    if acknowledged is not None:
        filters.append(Alert.acknowledged == acknowledged)
    if resolved is not None:
        filters.append(Alert.resolved == resolved)
    if severity is not None:
        filters.append(Alert.severity == severity)
    
    if filters:
        query = query.where(and_(*filters))
    
    # Order by newest first
    query = query.order_by(desc(Alert.created_at)).limit(limit)
    
    result = await db.execute(query)
    rows = result.all()
    
    # Build response with heat exchanger name
    alerts = []
    for alert, he_name in rows:
        alert_dict = {
            "id": alert.id,
            "heat_exchanger_id": alert.heat_exchanger_id,
            "heat_exchanger_name": he_name,
            "type": alert.type,
            "severity": alert.severity,
            "title": alert.title,
            "description": alert.description,
            "pump_id": alert.pump_id,
            "pump_name": alert.pump_name,
            "flow_rate": alert.flow_rate,
            "threshold": alert.threshold,
            "acknowledged": alert.acknowledged,
            "resolved": alert.resolved,
            "acknowledged_by": alert.acknowledged_by,
            "resolved_by": alert.resolved_by,
            "comments": alert.comments,
            "created_at": alert.created_at,
            "acknowledged_at": alert.acknowledged_at,
            "resolved_at": alert.resolved_at
        }
        alerts.append(AlertResponse(**alert_dict))
    
    return alerts


@router.get("/count")
async def get_alert_count(
    acknowledged: Optional[bool] = Query(None),
    resolved: Optional[bool] = Query(None),
    db: AsyncSession = Depends(get_session),
    current_user: User = Depends(get_current_user)
):
    """Get count of alerts matching filters"""
    query = select(Alert)
    
    filters = []
    if acknowledged is not None:
        filters.append(Alert.acknowledged == acknowledged)
    if resolved is not None:
        filters.append(Alert.resolved == resolved)
    
    if filters:
        query = query.where(and_(*filters))
    
    result = await db.execute(query)
    alerts = result.scalars().all()
    
    return {"count": len(alerts)}


@router.put("/{alert_id}/acknowledge")
async def acknowledge_alert(
    alert_id: int,
    data: AlertAcknowledge,
    db: AsyncSession = Depends(get_session),
    current_user: User = Depends(get_current_user)
):
    """Acknowledge an alert"""
    result = await db.execute(select(Alert).where(Alert.id == alert_id))
    alert = result.scalar_one_or_none()
    
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")
    
    if alert.acknowledged:
        raise HTTPException(status_code=400, detail="Alert already acknowledged")
    
    alert.acknowledged = True
    alert.acknowledged_by = current_user.username
    alert.acknowledged_at = datetime.utcnow()
    
    if data.comments:
        if alert.comments:
            alert.comments += f"\n\n[{datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')}] {current_user.username}: {data.comments}"
        else:
            alert.comments = f"[{datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')}] {current_user.username}: {data.comments}"
    
    await db.commit()
    
    return {"message": "Alert acknowledged", "alert_id": alert_id}


@router.put("/{alert_id}/resolve")
async def resolve_alert(
    alert_id: int,
    data: AlertResolve,
    db: AsyncSession = Depends(get_session),
    current_user: User = Depends(get_current_user)
):
    """Resolve an alert"""
    result = await db.execute(select(Alert).where(Alert.id == alert_id))
    alert = result.scalar_one_or_none()
    
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")
    
    if alert.resolved:
        raise HTTPException(status_code=400, detail="Alert already resolved")
    
    # Auto-acknowledge if not already acknowledged
    if not alert.acknowledged:
        alert.acknowledged = True
        alert.acknowledged_by = current_user.username
        alert.acknowledged_at = datetime.utcnow()
    
    alert.resolved = True
    alert.resolved_by = current_user.username
    alert.resolved_at = datetime.utcnow()
    
    if data.comments:
        if alert.comments:
            alert.comments += f"\n\n[{datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')}] {current_user.username} (RESOLVED): {data.comments}"
        else:
            alert.comments = f"[{datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')}] {current_user.username} (RESOLVED): {data.comments}"
    
    await db.commit()
    
    return {"message": "Alert resolved", "alert_id": alert_id}


@router.post("/{alert_id}/comment")
async def add_comment(
    alert_id: int,
    data: AlertComment,
    db: AsyncSession = Depends(get_session),
    current_user: User = Depends(get_current_user)
):
    """Add a comment to an alert"""
    result = await db.execute(select(Alert).where(Alert.id == alert_id))
    alert = result.scalar_one_or_none()
    
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")
    
    comment = f"[{datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')}] {current_user.username}: {data.comments}"
    
    if alert.comments:
        alert.comments += f"\n\n{comment}"
    else:
        alert.comments = comment
    
    await db.commit()
    
    return {"message": "Comment added", "alert_id": alert_id}


@router.delete("/heat-exchanger/{heat_exchanger_id}/clear-all")
async def clear_all_alerts(
    heat_exchanger_id: int,
    db: AsyncSession = Depends(get_session),
    current_user: User = Depends(get_current_user)
):
    """Clear (resolve) all alerts for a specific heat exchanger - Admin only"""
    # Check if user is admin
    if not current_user.is_admin:
        raise HTTPException(status_code=403, detail="Admin access required")
    
    # Verify heat exchanger exists
    result = await db.execute(
        select(HeatExchanger).where(HeatExchanger.id == heat_exchanger_id)
    )
    heat_exchanger = result.scalar_one_or_none()
    
    if not heat_exchanger:
        raise HTTPException(status_code=404, detail="Heat exchanger not found")
    
    # Get all unresolved alerts for this heat exchanger
    result = await db.execute(
        select(Alert).where(
            and_(
                Alert.heat_exchanger_id == heat_exchanger_id,
                Alert.resolved == False
            )
        )
    )
    alerts = result.scalars().all()
    
    # Mark all as resolved
    now = datetime.utcnow()
    count = 0
    for alert in alerts:
        alert.resolved = True
        alert.resolved_by = current_user.username
        alert.resolved_at = now
        if not alert.acknowledged:
            alert.acknowledged = True
            alert.acknowledged_by = current_user.username
            alert.acknowledged_at = now
        count += 1
    
    await db.commit()
    
    return {
        "message": f"Cleared {count} alert(s) for {heat_exchanger.name}",
        "count": count,
        "heat_exchanger_name": heat_exchanger.name
    }
